File: SAILSCOREWEB/sailmanager/frontend/src/lib/onlineEntryFields.ts
/**
 * Online entry field catalog (public form).
 * Source of truth for the admin panel and per-regatta configuration.
 */

export type OnlineEntrySectionId =
  | 'step1_class'
  | 'step2_helm'
  | 'step2_crew'
  | 'step3_boat'
  | 'payload_only';

export type OnlineEntryAppliesTo =
  | 'all'
  | 'one_design'
  | 'handicap'
  | 'multi_crew'; // sailors_per_boat >= 2 (crew step)

export interface OnlineEntryFieldDef {
  id: string;
  section: OnlineEntrySectionId;
  label: string;
  /** Key sent in POST /entries/ (or inside crew_members). */
  apiKey: string;
  appliesTo: OnlineEntryAppliesTo[];
  /** Shown on the public form. */
  inPublicForm: boolean;
  /** Required in HTML / frontend validation before submit. */
  requiredUi: boolean;
  /** Required on the backend (schema or business rule). */
  requiredBackend: boolean;
  /** Cannot be turned off in per-regatta config. */
  lockedCore: boolean;
  notes?: string;
}

export const ONLINE_ENTRY_SECTIONS: {
  id: OnlineEntrySectionId;
  title: string;
  step: number | null;
  description: string;
}[] = [
  {
    id: 'step1_class',
    title: 'Step 1 — Select class',
    step: 1,
    description: 'Class selection (One Design or Handicap).',
  },
  {
    id: 'step2_helm',
    title: 'Step 2 — Skipper details',
    step: 2,
    description: 'Skipper / helm details (always shown).',
  },
  {
    id: 'step2_crew',
    title: 'Step 2b — Crew members',
    step: 2,
    description: 'Only when the class has 2+ sailors (sailors_per_boat).',
  },
  {
    id: 'step3_boat',
    title: 'Step 3 — Boat details',
    step: 3,
    description: 'Boat fields; extra fields for Handicap vs One Design.',
  },
  {
    id: 'payload_only',
    title: 'Sent on submit (not a form field)',
    step: null,
    description: 'Derived values or payload-only; no dedicated input.',
  },
];

export const ONLINE_ENTRY_FIELDS: OnlineEntryFieldDef[] = [
  // —— Step 1 ——
  {
    id: 'class_name',
    section: 'step1_class',
    label: 'Competition class',
    apiKey: 'class_name',
    appliesTo: ['all'],
    inPublicForm: true,
    requiredUi: true,
    requiredBackend: true,
    lockedCore: true,
    notes: 'List comes from regatta classes (GET /regattas/{id}/classes/detailed).',
  },

  // —— Step 2 Helm ——
  {
    id: 'helm_position',
    section: 'step2_helm',
    label: 'Position (Skipper / Crew)',
    apiKey: 'helm_position',
    appliesTo: ['all'],
    inPublicForm: true,
    requiredUi: false,
    requiredBackend: false,
    lockedCore: false,
    notes: 'Default Skipper; stored as helm_position in payload.',
  },
  {
    id: 'first_name',
    section: 'step2_helm',
    label: 'First name',
    apiKey: 'first_name',
    appliesTo: ['all'],
    inPublicForm: true,
    requiredUi: true,
    requiredBackend: false,
    lockedCore: true,
  },
  {
    id: 'last_name',
    section: 'step2_helm',
    label: 'Last name',
    apiKey: 'last_name',
    appliesTo: ['all'],
    inPublicForm: true,
    requiredUi: true,
    requiredBackend: false,
    lockedCore: true,
  },
  {
    id: 'date_of_birth',
    section: 'step2_helm',
    label: 'Date of birth',
    apiKey: 'date_of_birth',
    appliesTo: ['all'],
    inPublicForm: true,
    requiredUi: false,
    requiredBackend: false,
    lockedCore: false,
  },
  {
    id: 'gender',
    section: 'step2_helm',
    label: 'Gender',
    apiKey: 'gender',
    appliesTo: ['all'],
    inPublicForm: true,
    requiredUi: true,
    requiredBackend: false,
    lockedCore: false,
  },
  {
    id: 'federation_license',
    section: 'step2_helm',
    label: 'Federation license',
    apiKey: 'federation_license',
    appliesTo: ['all'],
    inPublicForm: true,
    requiredUi: false,
    requiredBackend: false,
    lockedCore: false,
  },
  {
    id: 'email',
    section: 'step2_helm',
    label: 'Email',
    apiKey: 'email',
    appliesTo: ['all'],
    inPublicForm: true,
    requiredUi: true,
    requiredBackend: true,
    lockedCore: true,
    notes: 'Required to create sailor account and send confirmation emails.',
  },
  {
    id: 'contact_phone_1',
    section: 'step2_helm',
    label: 'Primary phone',
    apiKey: 'contact_phone_1',
    appliesTo: ['all'],
    inPublicForm: true,
    requiredUi: false,
    requiredBackend: false,
    lockedCore: false,
  },
  {
    id: 'contact_phone_2',
    section: 'step2_helm',
    label: 'Secondary phone',
    apiKey: 'contact_phone_2',
    appliesTo: ['all'],
    inPublicForm: true,
    requiredUi: false,
    requiredBackend: false,
    lockedCore: false,
  },
  {
    id: 'club',
    section: 'step2_helm',
    label: 'Club',
    apiKey: 'club',
    appliesTo: ['all'],
    inPublicForm: true,
    requiredUi: false,
    requiredBackend: false,
    lockedCore: false,
  },
  {
    id: 'helm_country',
    section: 'step2_helm',
    label: 'Country',
    apiKey: 'helm_country',
    appliesTo: ['all'],
    inPublicForm: true,
    requiredUi: false,
    requiredBackend: false,
    lockedCore: false,
  },
  {
    id: 'helm_country_secondary',
    section: 'step2_helm',
    label: 'Second country (dual nationality)',
    apiKey: 'helm_country_secondary',
    appliesTo: ['all'],
    inPublicForm: true,
    requiredUi: false,
    requiredBackend: false,
    lockedCore: false,
  },
  {
    id: 'territory',
    section: 'step2_helm',
    label: 'Territory / Federation',
    apiKey: 'territory',
    appliesTo: ['all'],
    inPublicForm: true,
    requiredUi: false,
    requiredBackend: false,
    lockedCore: false,
  },
  {
    id: 'address',
    section: 'step2_helm',
    label: 'Full address',
    apiKey: 'address',
    appliesTo: ['all'],
    inPublicForm: true,
    requiredUi: false,
    requiredBackend: false,
    lockedCore: false,
  },
  {
    id: 'zip_code',
    section: 'step2_helm',
    label: 'Postcode',
    apiKey: 'zip_code',
    appliesTo: ['all'],
    inPublicForm: true,
    requiredUi: false,
    requiredBackend: false,
    lockedCore: false,
  },
  {
    id: 'town',
    section: 'step2_helm',
    label: 'Town / City',
    apiKey: 'town',
    appliesTo: ['all'],
    inPublicForm: true,
    requiredUi: false,
    requiredBackend: false,
    lockedCore: false,
  },

  // —— Step 2b Crew (per member, crew_members[]) ——
  {
    id: 'crew_position',
    section: 'step2_crew',
    label: 'Position',
    apiKey: 'crew_members[].position',
    appliesTo: ['multi_crew'],
    inPublicForm: true,
    requiredUi: false,
    requiredBackend: false,
    lockedCore: false,
  },
  {
    id: 'crew_first_name',
    section: 'step2_crew',
    label: 'First name',
    apiKey: 'crew_members[].first_name',
    appliesTo: ['multi_crew'],
    inPublicForm: true,
    requiredUi: false,
    requiredBackend: false,
    lockedCore: false,
  },
  {
    id: 'crew_last_name',
    section: 'step2_crew',
    label: 'Last name',
    apiKey: 'crew_members[].last_name',
    appliesTo: ['multi_crew'],
    inPublicForm: true,
    requiredUi: false,
    requiredBackend: false,
    lockedCore: false,
  },
  {
    id: 'crew_email',
    section: 'step2_crew',
    label: 'Email',
    apiKey: 'crew_members[].email',
    appliesTo: ['multi_crew'],
    inPublicForm: true,
    requiredUi: false,
    requiredBackend: false,
    lockedCore: false,
    notes: 'Included in the array only if first name, last name, or email is filled.',
  },
  {
    id: 'crew_club',
    section: 'step2_crew',
    label: 'Club',
    apiKey: 'crew_members[].club',
    appliesTo: ['multi_crew'],
    inPublicForm: true,
    requiredUi: false,
    requiredBackend: false,
    lockedCore: false,
  },
  {
    id: 'crew_federation_license',
    section: 'step2_crew',
    label: 'Federation license',
    apiKey: 'crew_members[].federation_license',
    appliesTo: ['multi_crew'],
    inPublicForm: true,
    requiredUi: false,
    requiredBackend: false,
    lockedCore: false,
  },
  {
    id: 'crew_gender',
    section: 'step2_crew',
    label: 'Gender',
    apiKey: 'crew_members[].gender',
    appliesTo: ['multi_crew'],
    inPublicForm: true,
    requiredUi: false,
    requiredBackend: false,
    lockedCore: false,
  },
  {
    id: 'crew_helm_country',
    section: 'step2_crew',
    label: 'Country',
    apiKey: 'crew_members[].helm_country',
    appliesTo: ['multi_crew'],
    inPublicForm: true,
    requiredUi: false,
    requiredBackend: false,
    lockedCore: false,
    notes: 'Crew UI does not include phone, address, or extra fields beyond those listed.',
  },

  // —— Step 3 Boat ——
  {
    id: 'boat_name',
    section: 'step3_boat',
    label: 'Boat name',
    apiKey: 'boat_name',
    appliesTo: ['all'],
    inPublicForm: true,
    requiredUi: true,
    requiredBackend: false,
    lockedCore: true,
  },
  {
    id: 'boat_country_code',
    section: 'step3_boat',
    label: 'Country code (sail)',
    apiKey: 'boat_country_code',
    appliesTo: ['all'],
    inPublicForm: true,
    requiredUi: true,
    requiredBackend: true,
    lockedCore: true,
  },
  {
    id: 'sail_number',
    section: 'step3_boat',
    label: 'Sail number',
    apiKey: 'sail_number',
    appliesTo: ['all'],
    inPublicForm: true,
    requiredUi: true,
    requiredBackend: true,
    lockedCore: true,
    notes: 'Extra validation on submit (notify) in addition to HTML required.',
  },
  {
    id: 'boat_model',
    section: 'step3_boat',
    label: 'Boat model',
    apiKey: 'boat_model',
    appliesTo: ['handicap'],
    inPublicForm: true,
    requiredUi: false,
    requiredBackend: false,
    lockedCore: false,
  },
  {
    id: 'owner_first_name',
    section: 'step3_boat',
    label: 'Owner first name',
    apiKey: 'owner_first_name',
    appliesTo: ['handicap'],
    inPublicForm: true,
    requiredUi: false,
    requiredBackend: false,
    lockedCore: false,
  },
  {
    id: 'owner_last_name',
    section: 'step3_boat',
    label: 'Owner last name',
    apiKey: 'owner_last_name',
    appliesTo: ['handicap'],
    inPublicForm: true,
    requiredUi: false,
    requiredBackend: false,
    lockedCore: false,
  },
  {
    id: 'owner_email',
    section: 'step3_boat',
    label: 'Owner email',
    apiKey: 'owner_email',
    appliesTo: ['handicap'],
    inPublicForm: true,
    requiredUi: false,
    requiredBackend: false,
    lockedCore: false,
  },
  {
    id: 'category',
    section: 'step3_boat',
    label: 'Category (e.g. Men, Women, Mixed)',
    apiKey: 'category',
    appliesTo: ['one_design'],
    inPublicForm: true,
    requiredUi: false,
    requiredBackend: false,
    lockedCore: false,
    notes: 'Not sent for handicap (payload sets undefined).',
  },

  // —— Payload only ——
  {
    id: 'regatta_id',
    section: 'payload_only',
    label: 'Regatta ID',
    apiKey: 'regatta_id',
    appliesTo: ['all'],
    inPublicForm: false,
    requiredUi: false,
    requiredBackend: true,
    lockedCore: true,
  },
  {
    id: 'boat_country',
    section: 'payload_only',
    label: 'Boat country (duplicate of code)',
    apiKey: 'boat_country',
    appliesTo: ['all'],
    inPublicForm: false,
    requiredUi: false,
    requiredBackend: false,
    lockedCore: false,
    notes: 'Filled from boat_country_code on submit.',
  },
  {
    id: 'crew_members',
    section: 'payload_only',
    label: 'Crew members array',
    apiKey: 'crew_members',
    appliesTo: ['multi_crew'],
    inPublicForm: false,
    requiredUi: false,
    requiredBackend: false,
    lockedCore: false,
  },
];

/** EntryCreate schema fields not on the public online form. */
export const ONLINE_ENTRY_NOT_IN_PUBLIC_FORM: {
  apiKey: string;
  label: string;
  notes: string;
}[] = [
  { apiKey: 'bow_number', label: 'Bow number', notes: 'Admin / import only.' },
  { apiKey: 'rating', label: 'Rating (ANC)', notes: 'Handicap admin only; not on online entry.' },
  { apiKey: 'rating_type', label: 'Rating type', notes: 'orc | anc — admin.' },
  { apiKey: 'orc_low / orc_medium / orc_high', label: 'ORC ratings', notes: 'Admin only.' },
  { apiKey: 'paid', label: 'Paid', notes: 'Admin default false.' },
  { apiKey: 'confirmed', label: 'Confirmed', notes: 'Admin default false.' },
];

export function fieldsForSection(sectionId: OnlineEntrySectionId): OnlineEntryFieldDef[] {
  return ONLINE_ENTRY_FIELDS.filter((f) => f.section === sectionId);
}

export function countByAppliesTo(applies: OnlineEntryAppliesTo): number {
  return ONLINE_ENTRY_FIELDS.filter((f) => f.appliesTo.includes(applies)).length;
}

/** Default required for configurable fields (same as requiredUi in catalog). */
export function defaultRequiredForField(field: OnlineEntryFieldDef): boolean {
  return field.requiredUi;
}

export function defaultVisibleForField(field: OnlineEntryFieldDef): boolean {
  return field.inPublicForm;
}

export function isFieldConfigurable(field: OnlineEntryFieldDef): boolean {
  return field.inPublicForm && !field.lockedCore && field.section !== 'payload_only';
}

export const CONFIGURABLE_ONLINE_ENTRY_FIELDS = ONLINE_ENTRY_FIELDS.filter(isFieldConfigurable);

export function mergeEffectiveRequired(
  overrides?: Record<string, boolean> | null
): Record<string, boolean> {
  const out: Record<string, boolean> = {};
  for (const field of CONFIGURABLE_ONLINE_ENTRY_FIELDS) {
    out[field.id] =
      overrides && typeof overrides[field.id] === 'boolean'
        ? overrides[field.id]
        : defaultRequiredForField(field);
  }
  return out;
}

export function mergeEffectiveVisibility(
  overrides?: Record<string, boolean> | null
): Record<string, boolean> {
  const out: Record<string, boolean> = {};
  for (const field of CONFIGURABLE_ONLINE_ENTRY_FIELDS) {
    out[field.id] =
      overrides && typeof overrides[field.id] === 'boolean'
        ? overrides[field.id]
        : defaultVisibleForField(field);
  }
  return out;
}

export function computeOverridesFromEffective(
  effective: Record<string, boolean>
): Record<string, boolean> {
  const overrides: Record<string, boolean> = {};
  for (const field of CONFIGURABLE_ONLINE_ENTRY_FIELDS) {
    const def = defaultRequiredForField(field);
    if (effective[field.id] !== def) {
      overrides[field.id] = effective[field.id];
    }
  }
  return overrides;
}

export function computeVisibilityOverridesFromEffective(
  effective: Record<string, boolean>
): Record<string, boolean> {
  const overrides: Record<string, boolean> = {};
  for (const field of CONFIGURABLE_ONLINE_ENTRY_FIELDS) {
    const def = defaultVisibleForField(field);
    if (effective[field.id] !== def) {
      overrides[field.id] = effective[field.id];
    }
  }
  return overrides;
}

export function fieldAppliesToContext(
  field: OnlineEntryFieldDef,
  classType: 'one_design' | 'handicap' | undefined,
  multiCrew: boolean
): boolean {
  if (field.appliesTo.includes('all')) return true;
  if (field.appliesTo.includes('multi_crew') && multiCrew) return true;
  if (classType === 'handicap' && field.appliesTo.includes('handicap')) return true;
  if (classType === 'one_design' && field.appliesTo.includes('one_design')) return true;
  if (
    field.appliesTo.includes('one_design') &&
    !field.appliesTo.includes('handicap') &&
    classType !== 'handicap'
  ) {
    return true;
  }
  if (
    field.appliesTo.includes('handicap') &&
    !field.appliesTo.includes('one_design') &&
    classType === 'handicap'
  ) {
    return true;
  }
  return false;
}

export function isCoreRequiredField(fieldId: string): boolean {
  const field = ONLINE_ENTRY_FIELDS.find((f) => f.id === fieldId);
  return !!field?.lockedCore || !!field?.requiredBackend;
}

export function validateEntryAgainstRequired(
  payload: Record<string, unknown>,
  requiredMap: Record<string, boolean>,
  visibleMap: Record<string, boolean>,
  opts: {
    classType?: 'one_design' | 'handicap';
    multiCrew: boolean;
    crewMembers?: Record<string, unknown>[];
  }
): string | null {
  const { classType, multiCrew, crewMembers = [] } = opts;

  const check = (fieldId: string, value: unknown): string | null => {
    const field = ONLINE_ENTRY_FIELDS.find((f) => f.id === fieldId);
    if (!field || !requiredMap[fieldId]) return null;
    if (visibleMap[fieldId] === false) return null;
    if (!fieldAppliesToContext(field, classType, multiCrew)) return null;
    const filled =
      value != null && (typeof value !== 'string' || value.trim() !== '');
    if (!filled) return `${field.label} is required.`;
    return null;
  };

  for (const field of CONFIGURABLE_ONLINE_ENTRY_FIELDS) {
    if (field.section === 'step2_crew') continue;
    const apiKey = field.apiKey.replace('crew_members[].', '');
    const err = check(field.id, payload[apiKey]);
    if (err) return err;
  }

  const crewFieldMap: Record<string, string> = {
    crew_position: 'position',
    crew_first_name: 'first_name',
    crew_last_name: 'last_name',
    crew_email: 'email',
    crew_club: 'club',
    crew_federation_license: 'federation_license',
    crew_gender: 'gender',
    crew_helm_country: 'helm_country',
  };

  if (multiCrew) {
    for (const field of CONFIGURABLE_ONLINE_ENTRY_FIELDS.filter((f) => f.section === 'step2_crew')) {
      if (!requiredMap[field.id]) continue;
      const key = crewFieldMap[field.id];
      if (!key) continue;
      const activeMembers = crewMembers.filter(
        (m) =>
          (typeof m.first_name === 'string' && m.first_name.trim()) ||
          (typeof m.last_name === 'string' && m.last_name.trim()) ||
          (typeof m.email === 'string' && m.email.trim()) ||
          (typeof m.club === 'string' && m.club.trim()) ||
          (typeof m.federation_license === 'string' && m.federation_license.trim()) ||
          (typeof m.helm_country === 'string' && m.helm_country.trim())
      );
      if (activeMembers.length === 0) {
        return 'At least one crew member is required for this class.';
      }
      for (let i = 0; i < activeMembers.length; i++) {
        const err = check(field.id, activeMembers[i][key]);
        if (err) return `Crew member ${i + 1}: ${err}`;
      }
    }
  }

  return null;
}

export function createRequiredChecker(
  overrides?: Record<string, boolean> | null,
  classType?: 'one_design' | 'handicap',
  multiCrew = false
): (fieldId: string) => boolean {
  const requiredMap = mergeEffectiveRequired(overrides);
  return (fieldId: string) => {
    const field = ONLINE_ENTRY_FIELDS.find((f) => f.id === fieldId);
    if (!field) return false;
    if (field.lockedCore || field.requiredBackend) return true;
    if (!fieldAppliesToContext(field, classType, multiCrew)) return false;
    return !!requiredMap[fieldId];
  };
}

export function createVisibleChecker(
  overrides?: Record<string, boolean> | null,
  classType?: 'one_design' | 'handicap',
  multiCrew = false
): (fieldId: string) => boolean {
  const visibleMap = mergeEffectiveVisibility(overrides);
  return (fieldId: string) => {
    const field = ONLINE_ENTRY_FIELDS.find((f) => f.id === fieldId);
    if (!field) return true;
    if (field.lockedCore || field.requiredBackend) return true;
    if (!fieldAppliesToContext(field, classType, multiCrew)) return false;
    return visibleMap[fieldId] !== false;
  };
}

File: SAILSCOREWEB/sailmanager/backend/app/services/online_entry_fields.py
"""
Online entry field catalog and per-regatta public form configuration.
Mirrors frontend `src/lib/onlineEntryFields.ts` for IDs/defaults.
"""
from __future__ import annotations

from typing import Any, Dict, List, Optional, Tuple

from fastapi import HTTPException

from app import models, schemas

# Fields that are always visible/required on public online entry.
LOCKED_CORE_FIELD_IDS = {
    "class_name",
    "first_name",
    "last_name",
    "email",
    "boat_name",
    "boat_country_code",
    "sail_number",
}

# Default required flag for configurable fields (matches requiredUi in frontend catalog).
_CONFIGURABLE_REQUIRED_DEFAULTS: Dict[str, bool] = {
    "helm_position": False,
    "date_of_birth": False,
    "gender": True,
    "federation_license": False,
    "contact_phone_1": False,
    "contact_phone_2": False,
    "club": False,
    "helm_country": False,
    "helm_country_secondary": False,
    "territory": False,
    "address": False,
    "zip_code": False,
    "town": False,
    "crew_position": False,
    "crew_first_name": False,
    "crew_last_name": False,
    "crew_email": False,
    "crew_club": False,
    "crew_federation_license": False,
    "crew_gender": False,
    "crew_helm_country": False,
    "boat_model": False,
    "owner_first_name": False,
    "owner_last_name": False,
    "owner_email": False,
    "category": False,
}

# Default visibility for configurable fields.
_CONFIGURABLE_VISIBLE_DEFAULTS: Dict[str, bool] = {
    field_id: True for field_id in _CONFIGURABLE_REQUIRED_DEFAULTS
}

# field_id -> EntryCreate attribute (helm / boat)
_HELM_ATTR = {
    "helm_position": "helm_position",
    "date_of_birth": "date_of_birth",
    "gender": "gender",
    "federation_license": "federation_license",
    "contact_phone_1": "contact_phone_1",
    "contact_phone_2": "contact_phone_2",
    "club": "club",
    "helm_country": "helm_country",
    "helm_country_secondary": "helm_country_secondary",
    "territory": "territory",
    "address": "address",
    "zip_code": "zip_code",
    "town": "town",
}

_BOAT_ATTR = {
    "boat_model": "boat_model",
    "owner_first_name": "owner_first_name",
    "owner_last_name": "owner_last_name",
    "owner_email": "owner_email",
    "category": "category",
}

_CREW_ATTR = {
    "crew_position": "position",
    "crew_first_name": "first_name",
    "crew_last_name": "last_name",
    "crew_email": "email",
    "crew_club": "club",
    "crew_federation_license": "federation_license",
    "crew_gender": "gender",
    "crew_helm_country": "helm_country",
}

_APPLIES_HANDICAP_ONLY = {"boat_model", "owner_first_name", "owner_last_name", "owner_email"}
_APPLIES_ONE_DESIGN_ONLY = {"category"}
_APPLIES_MULTI_CREW = {
    "crew_position",
    "crew_first_name",
    "crew_last_name",
    "crew_email",
    "crew_club",
    "crew_federation_license",
    "crew_gender",
    "crew_helm_country",
}


def _sanitize_bool_map(raw: Any, *, field_name: str, allowed_ids: set[str]) -> Dict[str, bool]:
    if raw is None:
        return {}
    if not isinstance(raw, dict):
        raise ValueError(f"{field_name} must be an object.")
    out: Dict[str, bool] = {}
    for key, val in raw.items():
        if not isinstance(key, str):
            continue
        fid = key.strip()
        if fid not in allowed_ids:
            continue
        if not isinstance(val, bool):
            raise ValueError(f"Field '{fid}' must be true or false.")
        out[fid] = val
    return out


def sanitize_required_overrides(raw: Any) -> Dict[str, bool]:
    return _sanitize_bool_map(
        raw,
        field_name="online_entry_field_required",
        allowed_ids=set(_CONFIGURABLE_REQUIRED_DEFAULTS),
    )


def sanitize_visibility_overrides(raw: Any) -> Dict[str, bool]:
    return _sanitize_bool_map(
        raw,
        field_name="online_entry_field_visibility",
        allowed_ids=set(_CONFIGURABLE_VISIBLE_DEFAULTS),
    )


# Backward compatibility for older imports.
def sanitize_overrides(raw: Any) -> Dict[str, bool]:
    return sanitize_required_overrides(raw)


def merge_effective_required(overrides: Optional[Dict[str, bool]]) -> Dict[str, bool]:
    merged = dict(_CONFIGURABLE_REQUIRED_DEFAULTS)
    if overrides:
        for key, val in overrides.items():
            if key in merged and isinstance(val, bool):
                merged[key] = val
    return merged


def merge_effective_visibility(overrides: Optional[Dict[str, bool]]) -> Dict[str, bool]:
    merged = dict(_CONFIGURABLE_VISIBLE_DEFAULTS)
    if overrides:
        for key, val in overrides.items():
            if key in merged and isinstance(val, bool):
                merged[key] = val
    return merged


def normalize_field_overrides(
    required_raw: Any,
    visibility_raw: Any,
) -> Tuple[Dict[str, bool], Dict[str, bool]]:
    """
    Returns compact override maps normalized against defaults.
    Also enforces: hidden fields cannot remain required.
    """
    required_eff = merge_effective_required(sanitize_required_overrides(required_raw))
    visibility_eff = merge_effective_visibility(sanitize_visibility_overrides(visibility_raw))

    for field_id, visible in visibility_eff.items():
        if not visible:
            required_eff[field_id] = False

    required_compact: Dict[str, bool] = {}
    for field_id, default in _CONFIGURABLE_REQUIRED_DEFAULTS.items():
        if required_eff.get(field_id, default) != default:
            required_compact[field_id] = required_eff[field_id]

    visibility_compact: Dict[str, bool] = {}
    for field_id, default in _CONFIGURABLE_VISIBLE_DEFAULTS.items():
        if visibility_eff.get(field_id, default) != default:
            visibility_compact[field_id] = visibility_eff[field_id]

    return required_compact, visibility_compact


def field_applies(
    field_id: str,
    *,
    class_type: str,
    multi_crew: bool,
) -> bool:
    if field_id in _APPLIES_HANDICAP_ONLY:
        return class_type == "handicap"
    if field_id in _APPLIES_ONE_DESIGN_ONLY:
        return class_type != "handicap"
    if field_id in _APPLIES_MULTI_CREW:
        return multi_crew
    return field_id in _CONFIGURABLE_REQUIRED_DEFAULTS


def _is_filled(value: Any) -> bool:
    if value is None:
        return False
    if isinstance(value, str):
        return bool(value.strip())
    return True


def _crew_member_has_data(member: dict) -> bool:
    for key in ("first_name", "last_name", "email", "club", "federation_license", "helm_country"):
        if _is_filled(member.get(key)):
            return True
    return False


def validate_online_entry_fields(
    entry: schemas.EntryCreate,
    regatta: models.Regatta,
    *,
    class_type: str,
    sailors_per_boat: int,
) -> None:
    required_map = merge_effective_required(
        sanitize_required_overrides(getattr(regatta, "online_entry_field_required", None))
    )
    visibility_map = merge_effective_visibility(
        sanitize_visibility_overrides(getattr(regatta, "online_entry_field_visibility", None))
    )
    multi_crew = sailors_per_boat >= 2
    errors: List[str] = []

    for field_id, is_required in required_map.items():
        if not is_required:
            continue
        if not visibility_map.get(field_id, True):
            continue
        if not field_applies(field_id, class_type=class_type, multi_crew=multi_crew):
            continue

        if field_id in _HELM_ATTR:
            attr = _HELM_ATTR[field_id]
            if not _is_filled(getattr(entry, attr, None)):
                errors.append(f"{attr.replace('_', ' ').title()} is required.")
        elif field_id in _BOAT_ATTR:
            attr = _BOAT_ATTR[field_id]
            if not _is_filled(getattr(entry, attr, None)):
                errors.append(f"{attr.replace('_', ' ').title()} is required.")
        elif field_id in _CREW_ATTR:
            crew_key = _CREW_ATTR[field_id]
            members = entry.crew_members or []
            if not members:
                errors.append("At least one crew member is required for this class.")
                break
            for i, member in enumerate(members, start=1):
                if not isinstance(member, dict):
                    continue
                if not _crew_member_has_data(member):
                    continue
                if not _is_filled(member.get(crew_key)):
                    errors.append(
                        f"Crew member #{i}: {crew_key.replace('_', ' ').title()} is required."
                    )

    if errors:
        raise HTTPException(status_code=400, detail=errors[0] if len(errors) == 1 else errors)


def resolve_class_context(
    db,
    regatta_id: int,
    class_name: str,
) -> Tuple[str, int]:
    row = (
        db.query(models.RegattaClass)
        .filter(
            models.RegattaClass.regatta_id == regatta_id,
            models.RegattaClass.class_name == class_name,
        )
        .first()
    )
    if row:
        ct = (row.class_type or "one_design").strip().lower()
        spb = int(row.sailors_per_boat or 1)
        return ("handicap" if ct == "handicap" else "one_design", max(1, spb))
    return "one_design", 1
